### PR TITLE
metamask: only detectEthereumProvider() once

### PR DESCRIPTION
**Pull request by mrnerdhair**
This follows the style from the new `hdwallet-xdefi` and makes fixing https://github.com/shapeshift/hdwallet/issues/416 easier.

### CODE REVIEW

File: integration/src/integration.ts [hunk 1]
```diff
@@ -21,10 +21,6 @@ import { terraTests } from "./terra";
 import { kavaTests } from "./kava";
 import { WalletSuite } from "./wallets/suite";
 
-import { ethereum } from "./wallets/mocks/@metamask/detect-provider";
-
-jest.mock("@metamask/detect-provider", () => async () => Promise.resolve(ethereum));
-
```
Comment (mrnerdhair) [inline integration/src/integration.ts:27]: This was mocking `@metamask/detect-provider` during every wallet's integration test, not just MM.

File: integration/src/wallets/__mocks__/@metamask/detect-provider.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        case "eth_sendTransaction":
+          const [{ to }] = params;
+
+          return `txHash-${to}`;
+        default:
+          throw new Error(`ethereum: Unkown method ${method}`);
+      }
+    })
+  }
+}
```
Comment (mrnerdhair) [inline integration/src/wallets/__mocks__/@metamask/detect-provider.ts:24]: This is moved into the correct directory for jest's automock to work.

File: integration/src/wallets/metamask.ts [hunk 3]
```diff
@@ -10,8 +12,9 @@ export function createInfo(): core.HDWalletInfo {
 }
 
 export async function createWallet(): Promise<core.HDWallet> {
-  const wallet = new metamask.MetaMaskHDWallet();
-  await wallet.initialize();
+  const keyring = new core.Keyring();
+  const adapter = metamask.MetaMaskAdapter.useKeyring(keyring);
+  const wallet = await adapter.pairDevice()
```
Comment (mrnerdhair) [inline integration/src/wallets/metamask.ts:17]: Creating the wallet via the adapter (which is the public interface) now causes the provider to be pulled in from the `@metamask/detect-provider` mock.

File: packages/hdwallet-metamask/src/metamask.ts [hunk 4]
```diff
@@ -77,14 +76,11 @@ export class MetaMaskHDWallet implements core.HDWallet, core.ETHWallet {
...
-    try {
-      this.provider = await detectEthereumProvider({ mustBeMetaMask: true, silent: false, timeout: 3000 });
-    } catch (e) {
-      console.error(e);
-    }
-
-    return Promise.resolve();
+  public initialize(): never;
+  public initialize(provider: unknown): Promise<any>;
+  public async initialize(provider?: unknown): Promise<any> {
```
Comment (pastaghost) [inline packages/hdwallet-metamask/src/metamask.ts:81]: If `provider` is a required argument, what's the motivation to mark it as optional in the function definition?

**Review by pastaghost — APPROVED** [review]

File: packages/hdwallet-metamask/src/metamask.ts (hunk 4 above)
Comment (mrnerdhair) [inline packages/hdwallet-metamask/src/metamask.ts:81]:
```suggestion
  public async initialize(provider?: unknown): Promise<any> {
    if (this.provider) return;
```

**Comment by mrnerdhair** [comment]
@cjthompson @pastaghost I've updated the implementation here based on chris's feedback that web's using `wallet.initialize()`. Also updated #423 to harmonize xdefi with this approach: https://github.com/shapeshift/hdwallet/pull/423/commits/cad67f970373cd435e7eb2249da2852e75da3eb4

**Review by cjthompson — APPROVED** [review]

**Pull request merged**